README-dev.md

```diff
@@ -0,0 +1,170 @@
+the VSCode Extension API.
+
+#### Compile Code
+
+At this point it should be possible to execute
+```
+npm run compile
+```
+without any errors.
+
+_If there are errors_ something bad happened.
+
+## Development and Debugging
+
+### Development Branch
+
+Always work on a development branch, not the main branch:
+```
+git checkout -b <branch-name>
+```
+
+### Development
+
+Make whatever changes seem appropriate on your development branch.
+Use VSCode in order to get the debugger support for VSCode extension development.
+
+Compilation of the extension code is not done automatically by VSCode.
+It can be done manually by hitting `<ctrl>-<shift>-B` and choosing
+an appropriate task.
+
+The recommended task is `npm:watch` which will start a process that
+will automatically compile and build anytime a file changes.
+The process will run forever in the VSCode Terminal window once started.
+
+The same thing can be done in a shell window by executing
+```
+npm run watch
+```
+
+### Debugging
+
+The simplest way to launch the debugger is with the `F5` key.
+This should bring up a VSCode Extension Development Host -- another VSCode window with the modified Alive extension.
+Test the extension within this window.
+
+Standard VSCode debugging features (e.g. breakpoints, stepping through code) are available in the development VSCode window.
+A floating button bar on the upper right has icons for
+stepping through code and a red square outline
+which will kill the testing window.
+
+You can also start the debugger from the `Run and Debug` panel
+that you can start from the Activity Bar on the left.
+When this is showing there will be a small green triangle outline at the top of the panel which launch the Extension Development Host
+the same as the `F5` key.
+
+#### Output Panel
+
+In addition to the more sophisticated debugging features available from VSCode
+it is possible to view various types of output data.
+
+Show the `Output` view at the bottom of the screen (`<ctrl-K><ctrl-H>` or
+the **Output: Focus on Output View** command) or just select the `Output` tab on the bottom panel.
+On the right part of the title bar there is a dropdown to choose output from different threads.
+
+For Alive development the following are available (from the extension host only):
+
+* `Alive LSP` shows the output of the SBCL process running the Alive-lsp server.
+* `Alive Log` shows the typescript logging statements.
+
+#### Log Statements
+
+In order to send log statements to the `Alive Log` it is necessary to use the logging
+mechanism defined in `vscode/Log.ts`.
+_The output of `console.log()` calls is lost_.
+
+Import the necessary package:
+```
+import { log, toLog } from './vscode/Log'
+```
+where the specific relative path varies depending on the calling file.
+
+Log statements are very simple:
+```
+console.log("Hello world!")
+```
+Insert variable(s) into the log statement using typescript template strings:
+```
+console.log(`Failed to init ASDF tree: ${err}`)
```

Comment:
log(`Failed to init ASDF tree: ${toLog(err)}`)

Reply:
D'oh! Fixed.